### PR TITLE
E UI 4824 search location seperation another user scenario

### PR DESCRIPTION
**Pull request by waynepHMCTS**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/EUI-4824

### Change description ###

Compartmentise the search location component extract hearing specific code.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

**Review by johnbenjamin-hmcts — APPROVED** [review]

**Review by Kartheek-S — APPROVED** [review]

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.html [hunk 1]
```diff
@@ -1,12 +1,32 @@
...
     <main id="content" role="main">
-        <div *ngIf="serviceIds" class="search-location">
-          <exui-search-location class="search-location"
-                                [serviceIds]="serviceIds"
-                                [locationType]="'hearing'">
-          </exui-search-location>
+      <div class="govuk-form-group">
+        <div class="location-picker-custom">
+          <div class="search-location">
+            <label id="input-selected-location-label" class="govuk-body" for="input-selected-location">
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.html:7]: input-selected-location is not found now

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.html [hunk 2]
```diff
@@ -1,12 +1,32 @@
...
+                Search for a location by name
+            </label>
+            <br>
+            <exui-search-location class="search-location"
+              [selectedLocations$]="selectedLocations$"
+              [control]="findLocationFormGroup.controls.locationSelectedFormControl"
+              [disabled]=null
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.html:14]:
```suggestion
              [disabled]=false
```

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.spec.ts [hunk 3]
```diff
@@ -1,16 +1,33 @@
...
+class MockLocationSearchContainerComponent {
+  @Input() public serviceIds: string = '';
+  @Input() public locationType: string = '';
+  @Input() public disabled: boolean = false;
+  @Input() public selectedLocations$: Observable<LocationByEPIMSModel[]>;
+  @Input() public submitted?: boolean = true;
+  @Input() public control: AbstractControl;
+}
+
+fdescribe('LocationSearchContainerComponent', () => {
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.spec.ts:23]: fdescribe

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts [hunk 4]
```diff
@@ -1,16 +1,29 @@
 import { Component, OnInit } from '@angular/core';
-import { select, Store } from '@ngrx/store';
+import { FormBuilder, FormGroup } from '@angular/forms';
+import { Store, select } from '@ngrx/store';
+import { LocationModel } from 'api/locations/models/location.model';
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts:4]:
```suggestion
import { LocationByEPIMSModel } from '@hmcts/rpx-xui-common-lib/lib/models/location.model';
```

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts [hunk 5]
```diff
@@ -1,16 +1,29 @@
...
+import { Observable, of } from 'rxjs';
 import { map } from 'rxjs/operators';
 import * as fromHearingStore from '../../../../hearings/store';
-
 @Component({
   selector: 'exui-location-search-container',
   templateUrl: './location-search-container.component.html',
+  styleUrls: ['./location-search.component.scss']
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts:11]:
```suggestion
  styleUrls: ['./location-search-container.component.scss']
```

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts [hunk 6]
```diff
@@ -1,16 +1,29 @@
...
 })
 export class LocationSearchContainerComponent implements OnInit {
+  public locationType: string;
+  public findLocationFormGroup: FormGroup;
+  public selectedLocations$: Observable<LocationModel[]>;
+  public selectedLocation: LocationModel;
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts:17]:
```suggestion
  public selectedLocation: LocationByEPIMSModel ;
```

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts [hunk 7]
```diff
@@ -19,4 +32,18 @@ export class LocationSearchContainerComponent implements OnInit {
...
+  public addSelection(): void {
+    this.selectedLocations$.subscribe(selectedLocations => {
+        selectedLocations.push(this.findLocationFormGroup.controls.locationSelectedFormControl.value as LocationModel);
+        this.findLocationFormGroup.controls.locationSelectedFormControl.setValue(undefined);
+    });
+  }
+
+  public removeSelection(location: LocationModel): void {
+    this.selectedLocations$.subscribe(selectedLocations => {
+      const index = selectedLocations.findIndex(selectedLocation => selectedLocation.court_venue_id === location.court_venue_id);
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts:45]:
```suggestion
      const index = selectedLocations.findIndex(selectedLocation => selectedLocation.epims_id=== location.epims_id);
```

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts [hunk 8]
```diff
@@ -19,4 +32,18 @@ export class LocationSearchContainerComponent implements OnInit {
       this.serviceIds = id ? id : this.serviceIds;
     });
   }
+
+  public addSelection(): void {
+    this.selectedLocations$.subscribe(selectedLocations => {
+        selectedLocations.push(this.findLocationFormGroup.controls.locationSelectedFormControl.value as LocationModel);
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts:38]:
```suggestion
        selectedLocations.push(this.findLocationFormGroup.controls.locationSelectedFormControl.value as LocationByEPIMSModel);
```

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts (hunk 4 above)
Comment (rayliangatos) [reply src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts]: This is not resolved yet, we don't need LocationModel  anymore in angular, and it should refer to @hmcts/rpx-xui-common-lib/lib/models/location.model not api/locations/models/location.model

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts [hunk 9]
```diff
@@ -1,16 +1,29 @@
...
 @Component({
   selector: 'exui-location-search-container',
   templateUrl: './location-search-container.component.html',
+  styleUrls: ['./location-search-container.component.scss']
 })
 export class LocationSearchContainerComponent implements OnInit {
+  public locationType: string;
+  public findLocationFormGroup: FormGroup;
+  public selectedLocations$: Observable<LocationByEPIMSModel[]>;
+  public selectedLocation: LocationByEPIMSModel ;
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts:17]: empty space before ;

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts (hunk 4 above)
Comment (waynepHMCTS) [reply src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts]: Have you completed looking at this PR @rayliangatos. So I do not want to go backwards and forwards?

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts (hunk 6 above)
Comment (waynepHMCTS) [reply src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts]: Hi @rayliangatos this has been done

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts (hunk 4 above)
Comment (rayliangatos) [reply src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts]: I suppose I haven't asked you backwards changes, but if you don't want repeat comments please resovle the comments at a go.

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts (hunk 6 above)
Comment (rayliangatos) [reply src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts]: what I mean here is EMPTY_SPACE before ;

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.spec.ts [hunk 10]
```diff
@@ -1,16 +1,33 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
-import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
+import { Component, CUSTOM_ELEMENTS_SCHEMA, Input } from '@angular/core';
 import { provideMockStore } from '@ngrx/store/testing';
 import { LocationSearchContainerComponent } from './location-search-container.component';
+import { LocationByEPIMSModel, LocationModel } from '@hmcts/rpx-xui-common-lib/lib/models/location.model';
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.spec.ts:5]: unused LocationModel

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts [hunk 11]
```diff
@@ -1,16 +1,29 @@
 import { Component, OnInit } from '@angular/core';
-import { select, Store } from '@ngrx/store';
+import { FormBuilder, FormGroup } from '@angular/forms';
+import { Store, select } from '@ngrx/store';
+import { LocationByEPIMSModel, LocationModel } from 'api/locations/models/location.model';
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts:4]: comment is not resolved yet

**Review by rayliangatos — CHANGES_REQUESTED** [review]

File: src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts (hunk 11 above)
Comment (rayliangatos) [reply src/hearings/containers/request-hearing/location-search-container/location-search-container.component.ts:4]: could you check my suggested changes above?

**Review by rayliangatos — APPROVED** [review]

**Pull request merged**